pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,324 @@
+func (f *ftpFile) createJSONObjectReader() (datasource.RowReader, error) {
+	name := f.Name()
+
+	// Close the current file handle
+	if err := f.Close(); err != nil {
+		return nil, err
+	}
+
+	buffer := make([]byte, 1024)
+
+	// Retrieve the file as *ftp.Response
+	res, err := f.conn.Retr(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the response
+	n, err := res.Read(buffer)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer = buffer[:n]
+
+	reader := bytes.NewReader(buffer)
+
+	// Create a JSON decoder
+	decoder := json.NewDecoder(reader)
+
+	return &jsonReader{decoder: decoder}, nil
+}
+
+// createTextCSVReader creates a text reader for reading text files.
+func (f *ftpFile) createTextCSVReader() datasource.RowReader {
+	// Read the file content into a buffer
+	buffer := make([]byte, 1024)
+
+	n, err := f.response.Read(buffer)
+	if err != nil {
+		f.logger.Errorf("failed to read text file: %v", err)
+	}
+
+	buffer = buffer[:n]
+
+	reader := bytes.NewReader(buffer)
+
+	return &textReader{
+		scanner: bufio.NewScanner(reader),
+		logger:  f.logger,
+	}
+}
+
+// Next checks if there is another JSON object available.
+func (j jsonReader) Next() bool {
+	return j.decoder.More()
+}
+
+// Scan decodes the next JSON object into the provided structure.
+func (j jsonReader) Scan(i interface{}) error {
+	return j.decoder.Decode(&i)
+}
+
+// Next checks if there is another line available in the text file.
+func (f textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan scans the next line from the text file into the provided pointer to string.
+func (f textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
+		return nil
+	default:
+		return errNotPointer
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	// Close the FTP response.
+	err := f.response.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	// Retrieve the file content from FTP server and read into p
+	r, err := f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	f.offset = int64(len(b))
+
+	return len(b), nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	// Retrieve file content from the specified offset
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(resp)
+	if err != nil {
+		return 0, err
+	}
+
+	// Deep Copy
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	// setting current offset of the file
+	f.offset = off + int64(len(b))
```

Comment:
Please read the contract for `io.ReaderAt` and strictly comply with its provisions: https://pkg.go.dev/io@go1.22.5#ReaderAt

Specifically:
> If ReadAt is reading from an input source with a seek offset, ReadAt should not affect nor be affected by the underlying seek offset.

Reply:
Thanks for clarification.